File: packages/cli/src/targets/vscode/actions/spawn-process.js
// @ts-check
/// <reference path="./globals.js" />

/** @type {Map<string, import('child_process').ChildProcess>} */
const processRegistry = new Map()

/**
 * @param {{ command?: string, args?: string[], cwd?: string, env?: Record<string, string> } | null} payload
 * @param {(result: { processId?: string, pid?: number, error?: string }) => void} reply
 * @param {import('vscode').OutputChannel} channel
 * @param {import('vscode').Webview} webview
 * @returns {void}
 */
function spawnProcess(payload, reply, channel, webview) {
  const command = payload?.command ?? ''
  const args = payload?.args ?? []
  const options = {}

  if (payload?.cwd) options.cwd = payload.cwd
  if (payload?.env) options.env = { ...process.env, ...payload.env }

  const processId = 'proc_' + Math.random().toString(36).slice(2)

  try {
    const { spawn } = require('child_process')
    const proc = spawn(command, args, { ...options, stdio: ['pipe', 'pipe', 'pipe'] })

    if (!proc.pid) {
      reply({ error: `Failed to spawn process: ${command}` })
      return
    }

    processRegistry.set(processId, proc)
    reply({ processId, pid: proc.pid })

    proc.stdout.on('data', (chunk) => {
      webview.postMessage({
        processId,
        payload: { type: 'stdout', data: chunk.toString() },
      })
    })

    proc.stderr.on('data', (chunk) => {
      webview.postMessage({
        processId,
        payload: { type: 'stderr', data: chunk.toString() },
      })
    })

    proc.on('error', (err) => {
      webview.postMessage({
        processId,
        payload: { type: 'exit', exitCode: 1 },
      })
      processRegistry.delete(processId)
    })

    proc.on('exit', (code) => {
      webview.postMessage({
        processId,
        payload: { type: 'exit', exitCode: code ?? 1 },
      })
      processRegistry.delete(processId)
    })
  } catch (err) {
    reply({ error: err.message || 'Failed to spawn process' })
  }
}

File: packages/cli/src/targets/jetbrains/actions/SpawnProcess.kt
fun handleSpawnProcess(payload: org.json.JSONObject?, reply: org.json.JSONObject, browser: JBCefBrowser) {
    if (payload == null) {
        reply.put("type", "spawn-process:reply")
        val replyPayload = org.json.JSONObject()
        replyPayload.put("error", "Missing payload")
        reply.put("payload", replyPayload)
        return
    }

    val command = payload.optString("command", "")
    val args = payload.optJSONArray("args")?.let { arr ->
        (0 until arr.length()).map { arr.getString(it) }
    } ?: emptyList()
    val cwd = payload.optString("cwd", "")
    val env = payload.optJSONObject("env")

    val processId = "proc_" + java.util.UUID.randomUUID().toString().take(12)

    try {
        val cmdList = mutableListOf(command).apply { addAll(args) }
        val pb = ProcessBuilder(cmdList)
        if (cwd.isNotEmpty()) pb.directory(java.io.File(cwd))
        if (env != null) {
            val environment = pb.environment()
            for (key in env.keys()) {
                environment[key] = env.getString(key)
            }
        }

        val process = pb.start()
        jbProcessRegistry[processId] = process

        reply.put("type", "spawn-process:reply")
        val replyPayload = org.json.JSONObject()
        replyPayload.put("processId", processId)
        replyPayload.put("pid", process.pid())
        reply.put("payload", replyPayload)

        // Read stdout on background thread
        Thread {
            try {
                val reader = process.inputStream.bufferedReader()
                val buffer = CharArray(4096)
                var read: Int
                while (reader.read(buffer).also { read = it } != -1) {
                    val chunk = String(buffer, 0, read)
                    postStreamEvent(browser, processId, "stdout", chunk)
                }
            } catch (_: Exception) {}
        }.start()

        // Read stderr on background thread
        Thread {
            try {
                val reader = process.errorStream.bufferedReader()
                val buffer = CharArray(4096)
                var read: Int
                while (reader.read(buffer).also { read = it } != -1) {
                    val chunk = String(buffer, 0, read)
                    postStreamEvent(browser, processId, "stderr", chunk)
                }
            } catch (_: Exception) {}
        }.start()

        // Wait for exit on background thread
        Thread {
            val exitCode = process.waitFor()
            postStreamEvent(browser, processId, "exit", null, exitCode)
            jbProcessRegistry.remove(processId)
        }.start()

    } catch (e: Exception) {
        reply.put("type", "spawn-process:reply")
        val replyPayload = org.json.JSONObject()
        replyPayload.put("error", e.message ?: "Failed to spawn process")
        reply.put("payload", replyPayload)
    }
}
